src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -3733,5 +4003,163 @@ public override string ToString()
+//                {
+//                    totalClauses++;
+//                }
+//            }
+
+            builder.EmitCompressedUInt((uint)totalClauses);
+
+            for (int i = 0; i < _exceptionRegions.Length; i++)
+            {
+                ExceptionRegion exceptionRegion = _exceptionRegions[i];
+
+//                if (i > 0)
+//                {
+//                    ExceptionRegion previousClause = _exceptionRegions[i - 1];
+
+                    // If the previous clause has same try offset and length as the current clause,
+                    // but belongs to a different try block (CORINFO_EH_CLAUSE_SAMETRY is not set),
+                    // emit a special marker to allow runtime distinguish this case.
+                    //TODO: see above - do we need these
+//                    if ((previousClause.TryOffset == clause.TryOffset) &&
+//                        (previousClause.TryLength == clause.TryLength) &&
+//                        ((clause.Flags & CORINFO_EH_CLAUSE_FLAGS.CORINFO_EH_CLAUSE_SAMETRY) == 0))
+//                    {
+//                        builder.EmitCompressedUInt(0);
+//                        builder.EmitCompressedUInt((uint)RhEHClauseKind.RH_EH_CLAUSE_FAULT);
+//                        builder.EmitCompressedUInt(0);
+//                    }
+//                }
+
+                RhEHClauseKind clauseKind;
+
+                if (exceptionRegion.ILRegion.Kind == ILExceptionRegionKind.Fault ||
+                    exceptionRegion.ILRegion.Kind  == ILExceptionRegionKind.Finally)
+                {
+                    clauseKind = RhEHClauseKind.RH_EH_CLAUSE_FAULT;
+                }
+                else
+                if (exceptionRegion.ILRegion.Kind == ILExceptionRegionKind.Filter)
+                {
+                    clauseKind = RhEHClauseKind.RH_EH_CLAUSE_FILTER;
+                }
+                else
+                {
+                    clauseKind = RhEHClauseKind.RH_EH_CLAUSE_TYPED;
+                }
+
+                builder.EmitCompressedUInt((uint)exceptionRegion.ILRegion.TryOffset);
+
+                // clause.TryLength returned by the JIT is actually end offset... // TODO: does this apply to Wasm ExceptionRegion?
+                // https://github.com/dotnet/coreclr/issues/3585
+                int tryLength = (int)exceptionRegion.ILRegion.TryLength - (int)exceptionRegion.ILRegion.TryOffset;
+                builder.EmitCompressedUInt((uint)((tryLength << 2) | (int)clauseKind));
+
+                switch (clauseKind)
+                {
+                    case RhEHClauseKind.RH_EH_CLAUSE_TYPED:
+                        {
+                            builder.EmitCompressedUInt((uint)exceptionRegion.ILRegion.HandlerOffset);
+
+                            var type = (TypeDesc)_methodIL.GetObject((int)exceptionRegion.ILRegion.ClassToken); 
+
+                            // Once https://github.com/dotnet/corert/issues/3460 is done, this should be an assert.
```

Comment:
This can probably just be an assert (it will kick in for `catch (T)` in shared generic code).

Shared generic code is not enabled for WASM yet and once we enable it, we'll be able to solve it here easily.